src/main/java/net/fabricmc/loom/configuration/ide/RunConfig.java

```diff
@@ -157,29 +158,83 @@ public static RunConfig clientRunConfig(Project project) {
 		ideaClient.programArgs = "";
 		populate(project, extension, ideaClient, "client");
 		ideaClient.vmArgs += getOSClientJVMArgs();
-		ideaClient.vmArgs += " -Dfabric.dli.main=" + getMainClass("client", extension);
+		ideaClient.vmArgs += " -Dfabric.dli.main=" + getMainClass("client", extension, true);
 
 		return ideaClient;
 	}
 
+	@Deprecated // Replaced by runConfig(project, settings)
 	public static RunConfig serverRunConfig(Project project) {
 		LoomGradleExtension extension = project.getExtensions().getByType(LoomGradleExtension.class);
 
 		RunConfig ideaServer = new RunConfig();
 		ideaServer.configName = "Minecraft Server";
 		ideaServer.programArgs = "nogui ";
 		populate(project, extension, ideaServer, "server");
-		ideaServer.vmArgs += " -Dfabric.dli.main=" + getMainClass("server", extension);
+		ideaServer.vmArgs += " -Dfabric.dli.main=" + getMainClass("server", extension, false);
 
 		return ideaServer;
 	}
 
-	// This can be removed at somepoint, its not ideal but its the best solution I could thing of
+	public static RunConfig runConfig(Project project, LoomGradleExtension.RunConfigSettings settings) {
+		LoomGradleExtension extension = project.getExtensions().getByType(LoomGradleExtension.class);
+		String name = settings.getName();
+
+		boolean client = settings.isClient();
+
+		String configName = settings.getConfigName();
+		String mode = settings.getMode();
+
+		if (configName == null) {
+			configName = "Minecraft " + name.substring(0, 1).toUpperCase() + name.substring(1);
+		}
+
+		if (mode == null) {
+			mode = name;
+		}
+
+		RunConfig runConfig = new RunConfig();
+		runConfig.configName = configName;
+		populate(project, extension, runConfig, mode);
+
+		// Custom parameters
+		for (String progArg : settings.getProgramArgs()) {
+			runConfig.programArgs += " " + progArg;
+		}
+
+		for (String vmArg : settings.getVmArgs()) {
+			runConfig.vmArgs += " " + vmArg;
+		}
+
+		// Other mandatory properties
+		if (client) {
+			runConfig.vmArgs += getOSClientJVMArgs();
+		}
+
+		runConfig.vmArgs += " -Dfabric.dli.main=" + getMainClass(mode, extension, client);
+
+		// Remove unnecessary leading/trailing whitespaces we might have generated
+		runConfig.programArgs = runConfig.programArgs.trim();
+		runConfig.vmArgs = runConfig.vmArgs.trim();
+
+		return runConfig;
+	}
+
+	// This can be removed at somepoint, its not ideal but its the best solution I could think of
+	@Deprecated // Whatever this is still going to mean, replaced it with more strict version below
 	public static boolean needsUpgrade(File file) throws IOException {
 		String contents = FileUtils.readFileToString(file, StandardCharsets.UTF_8);
 		return !(contents.contains("net.fabricmc.devlaunchinjector.Main"));
 	}
```

Comment:
This can go now, its been long enough since this was added, it didnt really work anyway.

Reply:
Ok I'll remove the deprecated one. I'll keep `needsUpgrade(File, RunConfig)` below for now because it's still used and I'm not sure what to do with the usages. Should usages instead just assume an upgrade is always needed?